File: leetcode/python/Largest Rectangle in Histogram/Largest_Rectangle_in_Histogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height = [2,7,5,6,4]):
        n = len(height)
        if not height:
            return 0
        n = len(height)
        if n == 1:
            return height[0]
        stack = []
        max_area = 0
        for i in range(n+1):
            while stack and (i == n or height[stack[-1]] > height[i]):
                h = height[stack.pop()];
                if stack:
                    w = i - stack[-1] -1
                else:
                    w = i
                max_area = max(max_area,h * w)
            stack.append(i)
        return max_area

solu = Solution()
print solu.largestRectangleArea()

File: leetcode/cpp/LargestRectangleinHistogram.cpp
//
// Created by zhangqi on 2019/2/28.
//Time : O(n )
//Space : O(n)


#include <stack>
#include <vector>

using  namespace std;

class Solution {
public:
    /**
     * @param height: A list of integer
     * @return: The area of largest rectangle in the histogram
     */
    int largestRectangleArea(vector<int> &height) {

        int ans  = 0;
        stack<int> increaHegihtIndex;
        for(int i = 0; i <= height.size(); ){
            if(increaHegihtIndex.empty() ||   (i < height.size() && height[i] > height[increaHegihtIndex.top()])){
                increaHegihtIndex.push(i);
                i++;
            } else{
                int h = height[increaHegihtIndex.top()];
                increaHegihtIndex.pop();
                int left = increaHegihtIndex.empty() ?  -1 : increaHegihtIndex.top();
                ans = max( ans, h * (i - 1 - left )) ;
            }
        }
        return  ans;

    }
};